### PR TITLE
Extend FlaUI adapter

### PR DESCRIPTION
**Pull request by hieule-elca**

### CODE REVIEW

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        public static void LeftDoubleClickElement(this AutomationElement element)
+        {
+            var point = GetClickedPoint(element);
+            if (!point.IsEmpty)
+            {
+                AttachedMouse.LeftDoubleClick(point);               
+            }
+        }
+
+        public static Point GetClickedPoint(this AutomationElement element)
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs]: Is this name from White? If not, it should be `GetClickablePoint` as nothing was cllicked yet.

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                element.WaitUntilInputIsProcessed();
+            }
+        }
+
+        public static void RightClickElement(this AutomationElement element)
+        {
+            var point = GetClickedPoint(element);
+            if (!point.IsEmpty)
+            {
+                AttachedMouse.RightClick(point);
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs:40]: Any reason the `FocusElement` and `WaitUntilInputIsProcessed` is not added here as in `ClickElement`? It behaves differently and might cause confusion.

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            return Point.Empty;
+        }
+
+        public static void WaitUntilInputIsProcessed(this AutomationElement element)
+        {
+            Wait.UntilInputIsProcessed(TimeSpan.FromMilliseconds(500));            
+            var clickedPoint = element.TryGetClickedPoint();
+            var clickedPointX = !clickedPoint.IsEmpty ? clickedPoint.X : 0;
+            var clickedPointY = !clickedPoint.IsEmpty ? clickedPoint.Y : 0;
+            Retry.WhileTrue(() => AttachedMouse.IsCursorShowing(clickedPointX, clickedPointY), TimeSpan.FromMinutes(15));
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs:138]: Waiting for 15 Minutes? Is that how White behaves or just a number that fits? Also this method is more like `WaitWhileLoadingCursor` or something.

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.Common.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            var isEnabled = false;
+            Retry.WhileFalse(() =>
+            {
+                element.Properties.IsEnabled.TryGetValue(out isEnabled);
+                return isEnabled;
+            }, timeout: TimeSpan.FromSeconds(60), throwOnTimeout: false);
+            return isEnabled;
+        }
+
+        public static string ErrorProviderMessage(this AutomationElement element, Window window)
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Extensions/AdapterExtensions.Common.cs:196]: What is the use case of this method? Looks to be something very specific.

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ScrollableElement.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            if (verticalScrollBar == null)
+            {
+                return null;
+            }           
+            return verticalScrollBar.Get(SearchCriteria.ByText("Position"));
+        }
+
+        private static AutomationElement GetVerticalScrollBar(this AutomationElement element)
+        {
+            return element.Get(SearchCriteria.ByControlType(ControlType.ScrollBar).AndByText("Vertical Scroll Bar"))
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ScrollableElement.cs]: This probably does not work in localized Windows. Maybe use the values from https://github.com/FlaUI/FlaUI/blob/master/src/FlaUI.Core/Tools/LocalizedStrings.cs ?

File: src/FlaUI.Adapter.White/Mappings/ControlDictionaryItems.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            Add(ControlDictionaryItem.Primary(testControlType, controlType));
+        }
+
+        public virtual void AddSecondary(Type testControlType, ControlType controlType)
+        {
+            Add(ControlDictionaryItem.Secondary(testControlType, controlType));
+        }
+
+        public virtual void AddPrimary(Type testControlType, ControlType controlType, bool hasPrimaryChildren)
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/Mappings/ControlDictionaryItems.cs]: What is the boolean `hasPrimaryChildren` for?

File: src/FlaUI.Adapter.White/UIItems/WindowItems/Window.cs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            var framework = base.ModalWindows.FirstOrDefault(x => x.Name == title)?.FrameworkAutomationElement ??
+                this.GetChildren<AutomationElement>(SearchCriteria.ByControlType(ControlType.Window))
+                    .FirstOrDefault(x => x.Name == title)?.FrameworkAutomationElement;
+            return framework == null ? null : new Window(framework);
+        }              
+
+        public virtual void ReloadIfCached()
+        {
+            CacheRequest.ForceNoCache();
```
Comment (Roemer) [inline src/FlaUI.Adapter.White/UIItems/WindowItems/Window.cs]: This does not work but I also suspect that this is not really needed as caching was never really working great in White itself.

**Review by Roemer — COMMENTED** [review]
All in all I don't see anything bad. If you could look at the comments and reply or adjust accordingly that would be great.

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs (hunk 1 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs]: fixed

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs (hunk 2 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs:40]: fixed

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs (hunk 3 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ClickableElement.cs:138]: yes, it likes WaitWhileLoadingCursor, and we would like to wait until the waiting cursor is released.
It is not waitting for 15 minutes, it is just the timeout for Retry

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.Common.cs (hunk 4 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Extensions/AdapterExtensions.Common.cs:196]: it is migrated from TestStack.White, we will use it to get error message as I known

File: src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ScrollableElement.cs (hunk 5 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Extensions/AdapterExtensions.ScrollableElement.cs]: fixed

File: src/FlaUI.Adapter.White/Mappings/ControlDictionaryItems.cs (hunk 6 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/Mappings/ControlDictionaryItems.cs]: fixe

File: src/FlaUI.Adapter.White/UIItems/WindowItems/Window.cs (hunk 7 above)
Comment (hieule-elca) [reply src/FlaUI.Adapter.White/UIItems/WindowItems/Window.cs]: fixed

**Pull request merged**